perllib/FixMyStreet/App/Controller/Admin.pm

```diff
@@ -514,41 +636,165 @@ sub lookup_body : Private {
     return 1;
 }
 
+sub body_base : Chained('/') : PathPart('admin/body') : CaptureArgs(0) { }
+
 # This is for if the category name contains a '/'
-sub category_edit_all : Path('body') {
+sub category_edit_all : Chained('body_base') : PathPart('') {
     my ( $self, $c, $body_id, @category ) = @_;
     my $category = join( '/', @category );
-    $c->go( 'category_edit', [ $body_id, $category ] );
-}
 
-sub category_edit : Path('body') : Args(2) {
-    my ( $self, $c, $body_id, $category ) = @_;
+    $c->stash->{body_id} = $body_id;
+    $c->forward( 'lookup_body' );
 
+    my $contact = $c->stash->{body}->contacts->search( { category => $category } )->first;
+    $c->stash->{contact} = $contact;
+
+    $c->stash->{template} = 'admin/category_edit.html';
+    $c->forward( 'category_edit' );
+}
+
+sub body : Chained('body_base') : PathPart('') : CaptureArgs(1) {
+    my ( $self, $c, $body_id ) = @_;
     $c->stash->{body_id} = $body_id;
+}
+
+sub edit_body : Chained('body') : PathPart('') : Args(0) {
+    my ( $self, $c ) = @_;
+
+    unless ($c->user->has_permission_to('category_edit', $c->stash->{body_id})) {
+        $c->forward('check_for_super_user');
+    }
+
+    $c->forward( '/auth/get_csrf_token' );
+    $c->forward( 'lookup_body' );
+    $c->forward( 'fetch_all_bodies' );
+    $c->forward( 'body_form_dropdowns' );
+
+    if ( $c->get_param('posted') ) {
+        $c->log->debug( 'posted' );
+        $c->forward('update_contacts');
+        $c->forward('update_body_translations');
+    }
+
+    my @translations = $c->model('DB::Translation')->search(
+        {
+            tbl => 'body',
+            object_id => $c->stash->{body_id},
+            col => 'name'
+        }
+    )->all;
+
+    $c->stash->{translations} = \@translations;
+
+    $c->forward('fetch_languages');
+    $c->forward('fetch_contacts');
 
+    $c->stash->{template} = 'admin/body.html';
+    return 1;
+}
+
+sub body_translation_edit : Chained('body') : PathPart('translation') : Args(1) {
+    my ( $self, $c, $lang ) = @_;
+
+    $c->forward('check_for_super_user');
     $c->forward( '/auth/get_csrf_token' );
     $c->forward( 'lookup_body' );
 
+    my $translation = $c->model('DB::Translation')->find(
+        {
+            tbl => 'body',
+            object_id => $c->stash->{body_id},
+            col => 'name',
+            lang => $lang
+        }
+    );
+
+    $c->stash->{translation} = $translation;
+
+    my $language_name = '';
+    foreach my $supported_lang (@{$c->cobrand->languages}) {
+        my ($id, $name, $code) = split(',', $supported_lang);
+        $language_name = $name if $id eq $lang;
+    }
+
+    $c->stash->{language_name} = $language_name;
+
+    return 1;
+}
+
+sub category : Chained('body_base') : PathPart('') : CaptureArgs(2) {
```

Comment:
I like all the chained stuff. Should this be Chained off body rather than body_base?